File: lclang/src/main/java/lclang/types/Type.java
package lclang.types;

import lclang.libs.lang.classes.NullClass;

public abstract class Type implements Cloneable {
    public final String text;
    public boolean nullable = false;

    public Type(String text) {
        this.text = text;
    }

    public boolean isAccept(Type another) {
        if(NullClass.type.equals(this)) return true;

        if(another.isAccept(NullClass.type))
            return nullable;

        return isAcceptWithoutNullable(another) && (nullable || !another.nullable);
    }

    public boolean isAcceptWithoutNullable(Type another) {
        return false;
    }

    public Type nullable() {
        Type newType = clone();
        newType.nullable = true;
        return newType;
    }

    public Type notNullable() {
        Type newType = clone();
        newType.nullable = true;
        return newType;
    }

    @Override
    public String toString() {
        return (nullable ? "?": "") + text;
    }

    @Override
    public Type clone(){
        try {
            return (Type) super.clone();
        } catch (CloneNotSupportedException e) {
            throw new RuntimeException(e);
        }
    }
}

File: lclang/src/main/kotlin/lclang/types/Type.kt
package lclang.types

abstract class Type(val text: String): Cloneable {
    var nullable = false

    fun isAccept(another: Type): Boolean {
        if(Types.ANY==this || Types.UNDEFINED==another || another is Types.MagicType
            && another.text == text) return true
        return isAcceptWithoutNullable(another) && (nullable || !another.nullable)
    }

    open fun isAcceptWithoutNullable(another: Type): Boolean = false

    fun nullable(): Type {
        val newType = clone()
        newType.nullable = true
        return newType
    }

    fun notNullable(): Type {
        val newType = clone()
        newType.nullable = false
        return newType
    }

    override fun toString(): String {
        return (if(nullable) "?" else "") + text
    }

    public override fun clone(): Type {
        return super.clone() as Type
    }
}